agent/replacecall/src/main/java/randoop/instrument/ReplacementFileReader.java

```diff
@@ -0,0 +1,556 @@
+   * {@code ClassLoader} to find the replacement package.
+   *
+   * <p>Visits each class of the package on the classpath and applies {@link
+   * #addReplacementsForClass(ConcurrentHashMap, String, Class)} to add the method replacements.
+   *
+   * <p>Contrasts with {@link #addReplacementsForPackage(ConcurrentHashMap, String, String)} that
+   * searches for the package on the boot classpath.
+   *
+   * @param replacementMap the method replacement map to which new replacements are added
+   * @param originalPackage the original package name
+   * @param replacementPackage the replacement package name
+   * @param loader the {@code ClassLoader}
+   * @throws IOException if no package corresponding to replacement is found
+   * @throws ReplacementException if no replacements are found in the replacement package
+   * @see #addReplacementsForClassOrPackage(ConcurrentHashMap, String, String)
+   */
+  private static void addReplacementsForPackage(
+      ConcurrentHashMap<MethodSignature, MethodSignature> replacementMap,
+      String originalPackage,
+      String replacementPackage,
+      ClassLoader loader)
+      throws ReplacementException, IOException, ClassNotFoundException {
+    boolean found = false;
+    Enumeration<URL> resources;
+    resources = loader.getResources(replacementPackage.replace('.', '/'));
+
+    // Explore the whole classpath to ensure all replacements are found.
+    while (resources.hasMoreElements()) {
+      URL url = resources.nextElement();
+      try {
+        URLConnection connection = url.openConnection();
+        if (connection instanceof JarURLConnection) {
+          JarFile jarFile = ((JarURLConnection) connection).getJarFile();
+          addReplacementsFromAllClassesOfPackage(
+              replacementMap, originalPackage, replacementPackage, jarFile);
+        } else {
+          // The subclass for directories is internal.  It seems to work to assume the
+          // connection is a directory, and let an exception occur if it is not.
+          File path = new File(URLDecoder.decode(url.getPath(), "UTF-8"));
+          if (path.exists() && path.isDirectory()) {
+            addReplacementsForPackage(replacementMap, originalPackage, replacementPackage, path);
+            found = true;
+          }
+        }
+      } catch (IOException e) {
+        String msg =
+            String.format(
+                "Error identifying replacement %s with a package: %s",
+                replacementPackage, e.getMessage());
+        throw new ReplacementException(msg, e);
+      }
+    }
+    if (!found) {
+      String msg =
+          String.format(
+              "No package or class for replacement %s found on classpath", replacementPackage);
+      throw new ReplacementException(msg);
+    }
+  }
+
+  /**
+   * Adds method replacements for the classes in the replacement package to the replacement map.
+   *
+   * <p>Assumes that subpackage structure of original and replacement packages is identical.
```

Comment:
But individual files are allowed to be missing (from replacement only), right?  That would be helpful to clarify.